### PR TITLE
Add Namespace isolation

### LINKED ISSUE #792: Different deployment mode to support good performance for 10k Apps
**Describe the problem/challenge you have**
I need to use kapp-controller in a cluster to support deployment for 200 customers, and each customer has about 50 Apps. Totally we have more than 10000 Apps in the cluster. 

1. The queue is blocked because of so many apps
2.  More pods can't help because each pod shares the same queue, they will conflict.
3. If one customer start lots of App, other customers will wait, I hope each customer can has they own queue and quota.

**Describe the solution you'd like**
Add deployment mode support for kapp-controller, there are normal, cluster and namespace deployment mode support
1) Normal as current implementation
2) Cluster Model only reconcile the resource for cluster resource, like Package and PackageRepository
3) In Namespace deployment model, each deployment can define the work namespace environment parameter in deployment, and the deployment only reconcile the resource in the give namespace, like App and 
In this way, the namespace deployment will process the CRDs in its own namespace, and cluster deployment process the cluster CRDs

**Anything else you would like to add:**
There are some performance test result:
1) In Normal deployment, If there are 300 Apps is deployed, the reconcile will be very slow and new App will be triggered after more than 5m.  The deployment is to deploy to another Kubernetes cluster.
2) I did some change for CRD App to support namespace (https://github.com/showpune/carvel-kapp-controller/tree/cloud-version), we still can get good performance even there are 5000 apps in 100 namespaces
Current situation
image
Namespace deployment after change
image

There is similar issue for kpack and cartographer.

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

### PR DESCRIPTION
**Pull request by showpune**
#### What this PR does / why we need it:
Add two env RECONCILE_SCOPE and RECONCILE_NAMESPACE to control the reconcile scope
1) Add env RECONCILE_SCOPE to specify if reconcile only for cluster resource ( cluster ), namespace resource (namespace) or both (not set RECONCILE_SCOPE )
2) Add env RECONCILE_NAMESPACE to specify the reconcile namespace 
#### Which issue(s) this PR fixes:

Fixes #792

##### Review Checklist:

- [X] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [X] Relevant tests are added or updated
- [X] Relevant docs in this repo added or updated
- [ ] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [X] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

### CODE REVIEW

File: cmd/controller/run.go [hunk 1]
```diff
@@ -34,8 +34,10 @@ import (
 )
 
 const (
-	PprofListenAddr       = "0.0.0.0:6060"
-	kappctrlAPIPORTEnvKey = "KAPPCTRL_API_PORT"
+	PprofListenAddr             = "0.0.0.0:6060"
+	kappctrlAPIPORTEnvKey       = "KAPPCTRL_API_PORT"
+	kappctrlWorkNamespaceEnvKey = "WORK_NAMESPCE"
+	cluster                     = "cluster"
```
Comment (neil-hickey) [inline cmd/controller/run.go]: In the case where a namespaces is called "cluster" this won't work. This probably ought to be a separate flag instead of a "magic string" here.

In general we prefer flags over environment variables, this is true for the namespace stuff below also

File: cmd/controller/run.go [hunk 2]
```diff
@@ -59,6 +61,22 @@ func Run(opts Options, runLog logr.Logger) error {
 		restConfig.Timeout = opts.APIRequestTimeout
 	}
 
+	var workForClusterResource bool = true
```
Comment (neil-hickey) [inline cmd/controller/run.go]: workForClusterResource := true

File: cmd/controller/run.go [hunk 3]
```diff
@@ -59,6 +61,22 @@ func Run(opts Options, runLog logr.Logger) error {
...
+	var workForNamespaceResource bool = true
+	if namespaceWorkedFor, ok := os.LookupEnv(kappctrlWorkNamespaceEnvKey); ok {
+		opts.Namespace = namespaceWorkedFor
+		if namespaceWorkedFor == cluster {
+			workForClusterResource = true
```
Comment (neil-hickey) [inline cmd/controller/run.go]: The pattern of `workFor`... doesn't seem to fit quite right. Maybe some other options might be:

- ReconcileScope...
- Shard...
- ReconcileBy...
- FilterBy...

Any other opinions / options?

As mentioned above these names don't make things clear that you must have at least one `workForClusterResource` for the system to work (i.e the APIServer to be up and running to provide the CRDs to the cluster)

For example, the word singleton implies that at least one "cluster" thing is required

File: cmd/controller/run.go [hunk 4]
```diff
@@ -59,6 +61,22 @@ func Run(opts Options, runLog logr.Logger) error {
...
+			workForNamespaceResource = false
+		} else if namespaceWorkedFor != "" && namespaceWorkedFor != "None" {
```
Comment (neil-hickey) [inline cmd/controller/run.go]: This shouldn't be necessary: `namespaceWorkedFor != "None"`

This should be covered by having a flag instead.

File: cmd/controller/run.go [hunk 5]
```diff
@@ -214,7 +235,7 @@ func Run(opts Options, runLog logr.Logger) error {
 		}
 	}
 
-	{ // add controller for pkgrepositories
+	if workForClusterResource { // add controller for pkgrepositories
```
Comment (neil-hickey) [inline cmd/controller/run.go]: Is there a reason why PackageRepostories are not also handled like the other namespaced reconcilers?

File: config-namespace/0-namespace.yml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+#@ load("@ytt:data", "data")
+
+#! Prefix filename with "0-" so that generated config
+#! has Namespace first so that kubectl can can install kc
+#! (kapp of course perm automatic ordering)
+
+#@ if/end data.values.namespace != "default" and data.values.create_namespace:
```
Comment (neil-hickey) [inline config-namespace/0-namespace.yml]: Why is this file changed as part of this PR?

File: config-namespace/build.yml [hunk 7]
```diff
@@ -0,0 +1,23 @@
+#@ load("@ytt:data", "data")
```
Comment (neil-hickey) [inline config-namespace/build.yml]: Did you mean to commit this as a new file instead of changing the `config` folder?

File: config-namespace/deployment.yml [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+      - name: kapp-controller
+        image: kapp-controller
+        args:
+        - #@ "-enable-api-priority-and-fairness={}".format(data.values.enable_api_priority_and_fairness)
+        env:
+        - name: KAPPCTRL_MEM_TMP_DIR
+          value: /etc/kappctrl-mem-tmp
+        - name: KAPPCTRL_SIDECAREXEC_SOCK
+          value: /etc/kappctrl-mem-tmp/sidecarexec.sock
+        - name: WORK_NAMESPCE
```
Comment (neil-hickey) [inline config-namespace/deployment.yml]: It looks like you just added this WORK_NAMESPACE environment variable, but the whole file is new. Did you mean to change the `config/deployment.yml` file instead?

File: config-namespace/values.yml [hunk 9]
```diff
@@ -0,0 +1,17 @@
+#@data/values
```
Comment (neil-hickey) [inline config-namespace/values.yml]: Did you mean to commit this as a new file instead of changing the `config` folder?

File: config-namespace/rbac.yml [hunk 10]
```diff
@@ -0,0 +1,22 @@
+#@ load("@ytt:data", "data")
```
Comment (neil-hickey) [inline config-namespace/rbac.yml]: Did you mean to commit this as a new file instead of changing the `config` folder?

File: hack/deploy-namespace.sh [hunk 11]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
```
Comment (neil-hickey) [inline hack/deploy-namespace.sh]: This isn't needed for the PR.

File: config/rbac.yml [hunk 12]
```diff
@@ -67,6 +67,48 @@ rules:
 - apiGroups: ["flowcontrol.apiserver.k8s.io"]
   resources: ["prioritylevelconfigurations", "flowschemas"]
   verbs: ["list", "watch"]
+
+---
+kind: ClusterRole
```
Comment (neil-hickey) [inline config/rbac.yml:72]: This will cause a conflict when deploying multiple kapp-controllers as it is currently a Cluster-wide role.

This should be a `Role` with relevant `RoleBindings` per namespace.

**Review by neil-hickey — DISMISSED** [review]

File: cmd/controller/run.go (hunk 4 above)
Comment (showpune) [reply cmd/controller/run.go]: When I left empty in values.yaml, the value is "None" and I tried to keep compatible with it.

I can remove it

File: cmd/controller/run.go (hunk 5 above)
Comment (showpune) [reply cmd/controller/run.go]: I handle the PackageResository as a cluster resource (if workForClusterResource), what's your concern?

File: cmd/controller/run.go (hunk 3 above)
Comment (showpune) [reply cmd/controller/run.go]: Updated the key name

File: cmd/controller/run.go (hunk 2 above)
Comment (showpune) [reply cmd/controller/run.go]: Updated

File: cmd/controller/run.go (hunk 1 above)
Comment (showpune) [reply cmd/controller/run.go]: Use RECONCILE_NAMESPCE and RECONCILE_SCOPE to avoid magic string

File: config-namespace/0-namespace.yml (hunk 6 above)
Comment (showpune) [reply config-namespace/0-namespace.yml]: I use the config-namespace for namespaced deployment, I will remove it

File: config-namespace/build.yml (hunk 7 above)
Comment (showpune) [reply config-namespace/build.yml]: I will update it in config folder

File: config-namespace/deployment.yml (hunk 8 above)
Comment (showpune) [reply config-namespace/deployment.yml]: OK

File: config-namespace/rbac.yml (hunk 10 above)
Comment (showpune) [reply config-namespace/rbac.yml]: I agree. I planned to put all namespaced deployment in a seperate folder

File: config-namespace/values.yml (hunk 9 above)
Comment (showpune) [reply config-namespace/values.yml]: OK and I will remove the config-namespace folder

File: hack/deploy-namespace.sh (hunk 11 above)
Comment (showpune) [reply hack/deploy-namespace.sh]: I will remove it

File: config/rbac.yml (hunk 12 above)
Comment (showpune) [reply config/rbac.yml:72]: I think it RBAC to monitor to cluster resoource is needed. The clustorRole can be kept and we need to creat clusterrolebinding for the SA kapp-controller-sa in each namespace

File: cmd/controller/run.go (hunk 1 above)
Comment (joe-kimmel-vmw) [reply cmd/controller/run.go]: @showpune is it ok to put the 'A' back into `RECONCILE_NAMESPACE`(in 'space') ?   This looks like it was done consistently throughout so just checking whether there's a strong preference or it was a typo that got copy/pasted. thanks!

File: config/values.yml [hunk 13]
```diff
@@ -9,10 +9,12 @@ enable_api_priority_and_fairness: true
 
 dangerous_enable_pprof: false
 
-push_images: false
+push_images: true
```
Comment (joe-kimmel-vmw) [inline config/values.yml]: in the final submission please revert this `push_images` back to `false`

**Comment by joe-kimmel-vmw** [comment]
Another wrinkle with this feature as implemented is that as soon as I have made a kapp-controller that has namespace scope, now (a) it will introduce errors and race conditions if i start a kapp-controller also in "cluster" scope, and (b) it will require me to add one kapp-controller per namespace for any/every new namespace on this cluster. 

One thought I have is that we could do some regular expression matching (or regex-lite, ie just "*" wildcard) but that introduces other complexities as well.

File: config/build.yml [hunk 14]
```diff
@@ -1,24 +1,24 @@
 #@ load("@ytt:data", "data")
 
 apiVersion: kbld.k14s.io/v1alpha1
-kind: Sources
+kind: Config
```
Comment (neil-hickey) [inline config/build.yml:4]: Is this file should be reverted back to the original for it be mergable / tests to run

File: cmd/controller/run.go [hunk 15]
```diff
@@ -34,8 +34,12 @@ import (
...
 
 const (
-	PprofListenAddr       = "0.0.0.0:6060"
-	kappctrlAPIPORTEnvKey = "KAPPCTRL_API_PORT"
+	PprofListenAddr                  = "0.0.0.0:6060"
+	kappctrlAPIPORTEnvKey            = "KAPPCTRL_API_PORT"
+	kappctrlReconcileNamespaceEnvKey = "RECONCILE_NAMESPCE"
+	kappctrlReconcileScopeEnvKey     = "RECONCILE_SCOPE"
+	cluster                          = "cluster"
+	namespace                        = "namespace"
```
Comment (joe-kimmel-vmw) [inline cmd/controller/run.go]: per the comment I left earlier here is a slightly more fleshed out thought:
Here, rather than the "ReconcileScope" flag, we could introduce a boolean flag for "should i run the aggregate api server?" (yes/no) and an orthogonal environment variable for "what is the scope of the namespaces i'm reconciling?"    this second env var could use empty string ("") to indicate "all namespaces" or a legal namespace name and do an exact string match.  that way we could avoid regular expressions. :) 

so then the env vars would become something more like 
```
KAPPCTRL_START_API_SERVER
KAPPCTRL_RECONCILE_NAMESPACE
```

further, as we do prefer flags, and it turns out there's already a namespace flag: 
https://github.com/vmware-tanzu/carvel-kapp-controller/blob/8dba0472c9ebdca83d3dd981f783ff9afa75ac2b/cmd/controller/main.go#L24

then it's only necessary to add a boolean flag indicating whether or not to start the aggregate api server. 

Note that in this arrangement we would also be namespacing the PKGR resource - this was called out in a different comment but maybe now it's more clear why that's preferred, since we're slicing the functionality slightly differently.

Comment (showpune) [reply cmd/controller/run.go]: If we use KAPPCTRL_START_API_SERVER and KAPPCTRL_RECONCILE_NAMESPACE, it means we need to add another constant for namespace, see the sheet below:
image
Is it right?

For my point of view, reconcile scope and start api server just have indirect relationship. Is reconcile scope more straightforward to user?

Comment (joe-kimmel-vmw) [reply cmd/controller/run.go]: Hi @showpune - I don't think the above diagram makes sense, no.  I don't think we need that middle row, because I think all the reconcilers should be scoped. By scoping all the reconcilers together: app, pkgi, pkgr, pkg; all of those either are directed towards one namespace or all namespaces,  then the only question is "should this process also start the apiserver?"   

We definitely want to avoid magic strings, as for example in your example above a namespace called "None" would create problems.

If you're concerned that the process running the aggregate api server shouldn't also be reconciling anything at all, one option could be to point it at a namespace where we don't expect any resources to be created (such as the `kapp-controller` namespace of kapp-controller's deployment). 

Another option is to use regular expressions to specify which namespaces to match. This would introduce regular expressions with all the problems they bring, but it could also allow for an administrator to avoid needing to add an additional kapp-controller for every single namespace. (for example, maybe you'd want a single kapp-controller to reconcile all namespaces matching `dev-.*` to simplify developer onboarding?) 

I agree that reconcile scope and whether or not start the API server are unrelated. I think that's a point in favor of slicing the parameters this way, as it avoids having dependencies across flags.  As far as names of the flags, I think you would have better "user perspective" than I do, but we can also ask in our slack channel or community meeting.

does this make more sense, what I'm trying to propose?  What do you think?

Comment (showpune) [reply cmd/controller/run.go]: Hi @joe-kimmel-vmw，

It is OK that we use KAPPCTRL_START_API_SERVER and KAPPCTRL_RECONCILE_NAMESPACE. 

I want to confirm with you that if KAPPCTRL_START_API_SERVER is set to false, should we start reconciler for cluster resource like pkgr and pkg? The deployments are like below:
image
There should be a controller in each namespace, if we start reconciler in each namespace, if anything changed in cluster resource, it will be reconciled by all namespace with reconciler and there should be some data conflict issue

So the solution should be:
KAPPCTRL_START_API_SERVER=true: start the api server, reconcile the cluster resource like pkg and pkgr
KAPPCTRL_START_API_SERVER=false: not start the api server, not reconcile the cluster resource like pkg and pkgr

Is it right?

Comment (joe-kimmel-vmw) [reply cmd/controller/run.go]: @showpune I think there's still a misunderstanding. The way i see it, _all_ resources that are reconciled are namespaced. I'm not sure that we need the notion of a "cluster resource".  Am I missing something?

Comment (showpune) [reply cmd/controller/run.go]: Hi @joe-kimmel-vmw ,

I understood your idea and I commit the change, is it what you expect?

File: cmd/controller/run.go [hunk 16]
```diff
@@ -34,8 +34,10 @@ import (
 )
 
 const (
-	PprofListenAddr       = "0.0.0.0:6060"
-	kappctrlAPIPORTEnvKey = "KAPPCTRL_API_PORT"
+	PprofListenAddr                  = "0.0.0.0:6060"
+	kappctrlAPIPORTEnvKey            = "KAPPCTRL_API_PORT"
```
Comment (joe-kimmel-vmw) [inline cmd/controller/run.go]: I think we still prefer these to be flags rather than env vars, see e.g. :
https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/cmd/controller/main.go#L24

File: config/build.yml [hunk 17]
```diff
@@ -18,7 +18,7 @@ sources:
 #@ if/end data.values.push_images:
 ---
 apiVersion: kbld.k14s.io/v1alpha1
-kind: ImageDestinations
+kind: Config
```
Comment (joe-kimmel-vmw) [inline config/build.yml:21]: do we want this change?

File: cmd/controller/run.go [hunk 18]
```diff
@@ -59,6 +61,23 @@ func Run(opts Options, runLog logr.Logger) error {
...
+		if kappctrlStartApiServerString == "false" {
+			kappctrlStartApiServer = false
+		}
+		runLog.Info("Start with api server", "kappctrlStartApiServer", kappctrlStartApiServer)
+	} else {
+		runLog.Info("Start with api server with default value", "kappctrlStartApiServer", kappctrlStartApiServer)
+	}
+
+	if namespaceWorkedFor, ok := os.LookupEnv(kappctrlReconcileNamespaceEnvKey); ok {
+		opts.Namespace = namespaceWorkedFor
```
Comment (joe-kimmel-vmw) [inline cmd/controller/run.go]: i'm still curious whether you think there's a use-case for reconciling for multiple namespaces at all?   Do you think it would always be desirable to enumerate every single namespace and basically require a new instance of kapp-controller each time a namespace is added?  Although maybe the way you've implemented is great from a multi-tenancy isolation perspective.

**Review by joe-kimmel-vmw — COMMENTED** [review]
I suspect we also don't want you to delete [config-release/values.yml](https://github.com/vmware-tanzu/carvel-kapp-controller/pull/795/files#diff-09d86c4f70732b46a0ce953b97120419794501b14bb685325c3fe9766e52ae2d)

File: config/build.yml (hunk 17 above)
Comment (showpune) [reply config/build.yml:21]: I used this schema suggested by @neil-hickey 
image

and I checked the doc from https://carvel.dev/kbld/docs/v0.34.0/config/, is it the latest schema?

File: cmd/controller/run.go (hunk 18 above)
Comment (showpune) [reply cmd/controller/run.go]: For now, I want to one controller in one namespace for one customer instance, I can define the total resource quota for a customer in one namespace.  I prefer to use a new PR for multiple tenant support

Comment (joe-kimmel-vmw) [reply cmd/controller/run.go]: fair!

**Comment by joe-kimmel-vmw** [comment]
thanks @showpune -  Please also fix any remaining linter complaints and restore the file [config-release/values.yml](https://github.com/vmware-tanzu/carvel-kapp-controller/pull/795/files#diff-09d86c4f70732b46a0ce953b97120419794501b14bb685325c3fe9766e52ae2d). 

Then I also asked @cppforlife to take a look since he's the official approver for this project.

File: cmd/controller/main.go [hunk 19]
```diff
@@ -28,6 +28,7 @@ func main() {
 	flag.DurationVar(&ctrlOpts.APIRequestTimeout, "api-request-timeout", time.Duration(0), "HTTP timeout for Kubernetes API requests")
 	flag.BoolVar(&ctrlOpts.APIPriorityAndFairness, "enable-api-priority-and-fairness", true, "Enable/disable APIPriorityAndFairness feature gate for apiserver. Recommended to disable for <= k8s 1.19.")
 	flag.BoolVar(&sidecarexec, "sidecarexec", false, "Run sidecarexec")
+	flag.BoolVar(&ctrlOpts.StartAPIServer, "startAPIServer", true, "start apiserver")
```
Comment (cppforlife) [inline cmd/controller/main.go]: nit: startAPIServer => start-api-server

File: config/build.yml (hunk 17 above)
Comment (cppforlife) [reply config/build.yml:21]: i think it's fine to adopt `Config`

**Review by cppforlife — COMMENTED** [review]
im in favor of merging this PR once the flag casing it corrected.

File: cmd/controller/main.go (hunk 19 above)
Comment (showpune) [reply cmd/controller/main.go]: Done

**Comment by showpune** [comment]
create new PR https://github.com/vmware-tanzu/carvel-kapp-controller/pull/826
